packages/ifpack2/src/Ifpack2_BlockTriDiContainer_impl.hpp

```diff
@@ -527,16 +607,246 @@ namespace Ifpack2 {
+            Kokkos::parallel_for
+              (Kokkos::TeamThreadRange(member,num_vectors),[&](const local_ordinal_type &j) {
+                auto aptr = abase + blocksize_*(i + idiff*j);
+                auto bptr = &multivector_(blocksize_*lids_(i + ibeg_), j);
+                if (std::is_same<PackTag,ToBuffer>::value) 
+                  Kokkos::parallel_for
+                    (Kokkos::ThreadVectorRange(member,blocksize_),[&](const local_ordinal_type &k) {
+                      aptr[k] = bptr[k];
+                    });
+                else
+                  Kokkos::parallel_for
+                    (Kokkos::ThreadVectorRange(member,blocksize_),[&](const local_ordinal_type &k) {
+                      bptr[k] = aptr[k];
+                    });
+              });
+          });
+      } 
+      
+      void asyncSendRecvViaCudaStream(const impl_scalar_type_2d_view_tpetra &mv) {
+        IFPACK2_BLOCKTRIDICONTAINER_TIMER("BlockTriDi::AsyncableImport::AsyncSendRecv");
+
+#ifdef HAVE_IFPACK2_MPI
+        // constants and reallocate data buffers if necessary
+        const local_ordinal_type num_vectors = mv.extent(1);
+        const local_ordinal_type mv_blocksize = blocksize*num_vectors;
+
+        // 0. post receive async
+        for (local_ordinal_type i=0,iend=pids.recv.extent(0);i<iend;++i) {
+          irecv(comm, 
+                reinterpret_cast<char*>(buffer.recv.data() + offset_host.recv[i]*mv_blocksize),
+                (offset_host.recv[i+1] - offset_host.recv[i])*mv_blocksize*sizeof(impl_scalar_type),
+                pids.recv[i],
+                42,
+                &reqs.recv[i]);
+        }
+
+        // 1. async memcpy
+#if defined (KOKKOS_ENABLE_OPENMP)
+#pragma omp parallel for
+#endif        
+        for (local_ordinal_type i=0;i<static_cast<local_ordinal_type>(pids.send.extent(0));++i) {
+          auto &stream_at_i = stream.send[i];
+          // 1.0. enqueue pack buffer
+          copyViaCudaStream<ToBuffer>(lids.send, buffer.send, 
+                                      offset_host.send(i), offset_host.send(i+1), 
+                                      mv, blocksize,
+                                      stream_at_i);
+          
+          // // 1.1. enqueue call back posting isend , call back does not really work; what did I do wrong.
+          // auto &data_at_i = callback_data.send[i];
+          // data_at_i.comm = comm;
+          // data_at_i.buf = (char*)(buffer.send.data() + offset_host.send[i]*mv_blocksize);
+          // data_at_i.count = (offset_host.send[i+1] - offset_host.send[i])*mv_blocksize*sizeof(impl_scalar_type);
+          // data_at_i.pid = pids.send[i];
+          // data_at_i.tag = 42;
+          // data_at_i.ireq = &reqs.send[i];
+          // data_at_i.r_val = 0;
+          // CUDA_SAFE_CALL(cudaStreamAddCallback(stream_at_i, callback_isend, (void*)&data_at_i, 0));
+        }
+
+        Kokkos::fence();
+#if defined (KOKKOS_ENABLE_OPENMP)
+#pragma omp parallel for
```

Comment:
This parallel loop requires `MPI_THREAD_MULTIPLE`.  I worry that we don't have adequate testing for the case where MPI does not provide `MPI_THREAD_MULTIPLE`.

Reply:
This is just a pipelining experiment using cuda streams. I was working on a few things and cannot decouple each other but I put a macro so that the code path is not visible at this moment. So, we don't need to worry about this. In this PR, I do not use cuda stream (but some implementation is included with ifdef guards) nor use cuda space (but some implementation is already made to distinct tpetra memory space from internal memory space).

Reply:
@kyungjoo-kim cool, just wanted to make sure -- thanks!